File: components/browser_ui/modaldialog/android/java/src/org/chromium/components/browser_ui/modaldialog/ModalDialogFeatureMap.java
// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.browser_ui.modaldialog;

import android.content.Context;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.base.FeatureMap;
import org.chromium.build.annotations.NullMarked;
import org.chromium.components.cached_flags.CachedFlag;
import org.chromium.ui.base.DeviceFormFactor;

import java.util.List;

/** Java accessor for base::Features listed in {@link ModalDialogFeatureList}. */
@JNINamespace("browser_ui")
@NullMarked
public final class ModalDialogFeatureMap extends FeatureMap {
    private static final ModalDialogFeatureMap sInstance = new ModalDialogFeatureMap();

    public static final CachedFlag sDialogsOnLargeFormFactors =
            new CachedFlag(sInstance, ModalDialogFeatureList.DIALOGS_ON_LARGE_FORM_FACTORS, false);
    public static final CachedFlag sModalDialogLayoutWithSystemInsets =
            new CachedFlag(
                    sInstance, ModalDialogFeatureList.MODAL_DIALOG_LAYOUT_WITH_SYSTEM_INSETS, true);
    public static final List<CachedFlag> sCachedFlags =
            List.of(sDialogsOnLargeFormFactors, sModalDialogLayoutWithSystemInsets);

    /**
     * Returns whether large form factor modal dialog UI updates should be applied.
     *
     * @param context The {@link Context} associated with the window or activity.
     * @return True if the feature flag is enabled and the form factor is a large form factor.
     */
    public static boolean isLargeFormFactorUiEnabled(Context context) {
        return sDialogsOnLargeFormFactors.isEnabled()
                && DeviceFormFactor.isNonMultiDisplayContextOnTablet(context);
    }

    // Do not instantiate this class.
    private ModalDialogFeatureMap() {}

    /**
     * @return the singleton ModalDialogFeatureMap.
     */
    public static ModalDialogFeatureMap getInstance() {
        return sInstance;
    }

    /** Convenience method to call {@link #isEnabledInNative(String)} statically. */
    public static boolean isEnabled(String featureName) {
        return getInstance().isEnabledInNative(featureName);
    }

    @Override
    protected long getNativeMap() {
        return ModalDialogFeatureMapJni.get().getNativeMap();
    }

    @NativeMethods
    public interface Natives {

        long getNativeMap();
    }
}

File: components/browser_ui/modaldialog/android/modaldialog_feature_map.cc
// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/android/feature_map.h"
#include "base/no_destructor.h"
#include "components/browser_ui/modaldialog/android/features.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "components/browser_ui/modaldialog/android/modaldialog_jni_headers/ModalDialogFeatureMap_jni.h"

namespace browser_ui {

namespace {

// Array of features exposed through the Java ModalDialogFeatureMap API.
// Entries in this array may either refer to features defined in
// components/browser_ui/modaldialog/android/features.h or in other
// locations in the code base (e.g. content_features.h).
const base::Feature* const kFeaturesExposedToJava[] = {
    &kModalDialogLayoutWithSystemInsets,
};

// static
base::android::FeatureMap* GetFeatureMap() {
  static base::NoDestructor<base::android::FeatureMap> kFeatureMap(std::vector(
      std::begin(kFeaturesExposedToJava), std::end(kFeaturesExposedToJava)));
  return kFeatureMap.get();
}

}  // namespace

static jlong JNI_ModalDialogFeatureMap_GetNativeMap(JNIEnv* env) {
  return reinterpret_cast<jlong>(GetFeatureMap());
}

}  // namespace browser_ui
